js/walk.go

```diff
@@ -0,0 +1,259 @@
+package js
+
+// IVisitor represents the AST Visitor
+// Each INode encountered by `Walk` is passed to `Enter`, children nodes will be ignored if the returned IVisitor is nil
+// `Exit` is called upon the exit of a node
+type IVisitor interface {
+	Enter(n INode) IVisitor
+	Exit(n INode)
+}
+
+// Walk traverses an AST in depth-first order
+func Walk(v IVisitor, n INode) {
+	if n == nil {
+		return
+	}
+
+	// check if pointer is nil
+	if n.isNil() {
```

Comment:
This is the same as `n == nil`, as all implementations of `isNil` just check `n == nil`. Let's remove `isNil` entirely.

Reply:
`isNil`  (pointer receiver) is required since it gives us compile-time warnings when people try to pass a non-pointer to `Walk` . It also differentiates an `INode`  from any other interface.

Reply:
Is that really necessary? I don't check for nil pointers in any function where pointers are passed, because that is obviously a bug in some other part. Adding all these checks really slows down, and nil check are implicit in Go so there really is no need. I'd suggest to see this check and the `n == nil` check above gone.

Reply:
it is necessary actually, if we don't it will panic at some point during the Walk. Mainly because some struct fields are optional and can contain nil pointers:

https://github.com/karelorigin/parse/blob/master/js/ast.go#L431

Reply:
Oh! I didn't notice the double is nil check, yes, you're right. That doesn't appear to be necessary. I'll see if I can make the necessary changes.